### PR TITLE
$81 knockback

### PR DESCRIPTION
**Pull request by MrJackey**
## Summary
Fixes 81
Objects can now be knockbacked by spikes if they have the knockbackable component.
Also fixed the transfoms components of the player prefab.

## Visualisation
EBztMVYnpQ

### CODE REVIEW

File: Assets/Scripts/Common/DamageArea.cs [hunk 1]
```diff
@@ -39,32 +41,39 @@ private enum DamageMode {
...
 
+		CalculateAvgContact(other, out Vector2 averageContact);
+
 		Health healthComponent = other.attachedRigidbody.GetComponent<Health>();
 		if (healthComponent != null) {
 			cooldownTimer = damageCooldown;
 			Health health = healthComponent;
 			health.TakeDamage(isEnter ? damage : damage * Time.deltaTime);
 
 			if (other.attachedRigidbody.CompareTag(effectTarget))
```
Comment (smeas) [inline Assets/Scripts/Common/DamageArea.cs]:
```suggestion
			if (effectTarget.Length != 0 && other.attachedRigidbody.CompareTag(effectTarget))
```

File: Assets/Scripts/Common/DamageArea.cs [hunk 2]
```diff
@@ -39,32 +41,39 @@ private enum DamageMode {
...
+
+		Knockbackable knockbackComponent = other.attachedRigidbody.GetComponent<Knockbackable>();
+		if (knockbackComponent != null) {
+			Vector2 direction = (Vector2) other.transform.position - averageContact;
+			knockbackComponent.Knockback(direction.normalized, knockbackForce, knockbackDuration);
 		}
 	}
 
-	private void DoDamageEffect(Collider2D other) {
+	private void CalculateAvgContact(Collider2D other, out Vector2 avg) {
```
Comment (smeas) [inline Assets/Scripts/Common/DamageArea.cs]: Why is `avg` an out parameter, can't you just return it?

File: Assets/Scripts/Common/Knockbackable.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+	public void Knockback(Vector2 direction, float force, float duration) {
+		onKnockback.Invoke();
+		velocity = direction * force;
+
+		if (knockbackRoutine != null)
+			StopCoroutine(knockbackRoutine);
+
+		doKnockback = true;
+		knockbackRoutine = StartCoroutine(CoKnockbackDuration(duration));
```
Comment (smeas) [inline Assets/Scripts/Common/Knockbackable.cs]: `knockbackDurationRoutine` would be a better name.

**Review by smeas — CHANGES_REQUESTED** [review]

**Review by Addehman — COMMENTED** [review]
Is there anything forcing the player to be facing the same direction that the player was travelling in when entering a damageArea? Thinking this might be important for creating the feeling of being pushes/knocked backwards?

**Review by smeas — APPROVED** [review]
:+1:

**Pull request merged**